### ISSUE TITLE
[java] EmptyStatementNotInLoop reported False positive 

### ISSUE THREAD

**Issue by pagarwal-ignitetech** [pmd/pmd#2966]
**Affects 6.29.0** 

Rule: [EmptyStatementNotInLoop](https://pmd.github.io/latest/pmd_rules_java_errorprone.html#emptystatementnotinloop)

Tested in latest PMD.

The below sample code semi-colon just before comment is valid and shouldn't be detected by PMD.

```java
 if (widget.direction == Directions.TOP_LEFT ||
		widget.direction == Directions.TOP ||
		widget.direction == Directions.LEFT)
	      ; // someday gesture = CONTAINER_POSITIONING_GESTURE;
```

**Steps to reproduce:**

Run PMD CLI command for all rules (including EmptyStatementNotInLoop) and got above in report.

**Comment by oowekyala** [comment]
Can you just add `NOPMD` at the start of the comment? I think it's preferable to use suppression here than to add an exception to the rule.

**Comment by pagarwal-ignitetech** [comment]
> Can you just add `NOPMD` at the start of the comment? I think it's preferable to use suppression here than to add an exception to the rule.

This is our clients code, which we can't amend any other way to solve this.

**Comment by pagarwal-ignitetech** [comment]
This is real FP by PMD as you see 
```Java
if (a == true)
        ; else {
            
        }
```
removing semicolon just before else will lead to compilation error.

**Comment by adangel** [comment]
I think, PMD should still report this, as this is a if statement with a empty body. This is in any case code, that should be look at. Like it is now, the whole if statement can be removed, because it has no effect.

But we see here only one single snippet and not the whole class, so we don't have the context. Right now, I'm not convinced, that this is a bug/false positive in PMD.

**Comment by Fernal73** [comment]
There's this:

```java
for (int i=0;  i < argc;  i++)
{
   if (argv[i]=="left")
      hpos++;
   else if (argv[i]=="right")
      hpos--;
   else if (argv[i]=="up")
      ;
   else if (argv[i]=="down")
      ;
   else System.err.printf("Unknown option %s%n", argv[i]);

}
```
https://stackoverflow.com/a/16429324/3924108

Will a property for this scenario do?

**Comment by Fernal73** [comment]
> 
> This is our clients code, which we can't amend any other way to solve this.

This is no reason for PMD to change anything.

**Comment by Fernal73** [comment]
@pagarwal-ignitetech Any reason for closing this?

**Comment by pagarwal-ignitetech** [comment]
> @pagarwal-ignitetech Any reason for closing this?

I closed as above was declared as no bug , or false positive from PMD. And in this case there can be two solutions: either negate the if or if is empty {..} with comments and else part is there.

**Comment by Fernal73** [comment]
> > @pagarwal-ignitetech Any reason for closing this?
> 
> I closed as above was declared as no bug , or false positive from PMD. And in this case there can be two solutions: either negate the if or if is empty {..} with comments and else part is there.

I'm with allowing it if it makes the code more readable (as per the [SO example](https://github.com/pmd/pmd/issues/2966#issuecomment-743774361)) . But I'd rather refactor the code.

**Comment by Fernal73** [comment]
> > > @pagarwal-ignitetech Any reason for closing this?
> > 
> > 
> > I closed as above was declared as no bug , or false positive from PMD. And in this case there can be two solutions: either negate the if or if is empty {..} with comments and else part is there.
> 
> I'm with allowing it if it makes the code more readable (as per the [SO example](https://github.com/pmd/pmd/issues/2966#issuecomment-743774361)) . But I'd rather refactor the code.

Whether it's worth amending the rule or not depends on how common or uncommon this occurrence is.

**Comment by pagarwal-ignitetech** [comment]
@Fernal73 , Ofcourse semi colon with comment in one flow of if-else is confusing, i would refactor below 
```Java
if(condition) 
; // Note this
else {
//perform operations
}
``` as
```Java
if(condition) {
// Note this
} else {
//perform operations
}
```

**Comment by Fernal73** [comment]
> @Fernal73 , Ofcourse semi colon with comment in one flow of if-else is confusing, i would refactor below
> 
> ```java
> if(condition) 
> ; // Note this
> else {
> //perform operations
> }
> ```
> 
> as
> 
> ```java
> if(condition) {
> // Note this
> } else {
> //perform operations
> }
> ```

I was referring to the StackOverflow example.

**Comment by Fernal73** [comment]
@pagarwal-ignitetech Your sample is best suppressed as the rule stands.

**Comment by Fernal73** [comment]
> @pagarwal-ignitetech Your sample is best suppressed as the rule stands.

You can reopen this issue if you're not convinced.

**Comment by Fernal73** [comment]
> This is real FP by PMD as you see
> 
> ```java
> if (a == true)
>         ; else {
>             
>         }
> ```
> 
> removing semicolon just before else will lead to compilation error.

Checkstyle doesn't flag this either.

**Comment by Fernal73** [comment]
This is legitimate, as well.

```java
public enum Hello {
;

public static void main(String... args) {

}

```

**Comment by oowekyala** [comment]
@Fernal73 I don't think your stackoverflow example is something common or useful. It can be refactored by replacing the semi by an empty block like @pagarwal-ignitetech mentioned. So I don't think it's worth changing the rule. Please stop spamming one-liners on this issue.

**Comment by Fernal73** [comment]
> @Fernal73 I don't think your stackoverflow example is something common or useful. It can be refactored by replacing the semi by an empty block like @pagarwal-ignitetech mentioned. So I don't think it's worth changing the rule. Please stop spamming one-liners on this issue.

The StackOverflow example can be refactored via a switch (as far as I'm concerned). As for the original sample, it's worth considering adding a property simply because it isn't a dangling semi-colon and is syntactically correct. I don't see why one would refactor that to use an empty block instead. It's either best suppressed or a property added to account for this case (optionally, of course).
